### ISSUE TITLE
A Tuple of Literals cannot be used used as keys of a Dict

### ISSUE THREAD

**Issue by Congee** [microsoft/pyright#959]
**Describe the bug**
see code below

**To Reproduce**
see code below

**Expected behavior**
Pyright should not report an error to say the least.

**Screenshots or Code**
```python
from typing import Literal

a: Literal[0, 1] = ...
b: Literal[0] = ...  # type: ignore

{
    (0, 0): ...,
    (1, 0): ...,  # pyright reports error here
}[(a, b)]
```
```
Argument of type "Tuple[Literal[0, 1], Literal[0]]" cannot be assigned to parameter "k" of type "Tuple[Literal[0], Literal[0]] | Tuple[Literal[1], Literal[0]]" in function "__getitem__"
  Tuple entry 1 is incorrect type
  Tuple entry 1 is incorrect type
    Type "Literal[1]" cannot be assigned to type "Literal[0]"
      "Literal[1]" cannot be assigned to type "Literal[0]"
    Type "Literal[0]" cannot be assigned to type "Literal[1]"
...
[Pyright (reportGeneralTypeIssues):]
```

**VS Code extension or command-line**
vim + coc-pyright with pyright 1.1.64

**Additional context**
FWIW, I'm trying to do pattern matching and I don't expect pyright to supprot it :)

**Comment by erictraut** [comment]
The behavior you're seeing comes from the inference rules Pyright uses for [dict](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#dictionary-expressions) and [tuple](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#tuple-expressions) expressions. So the behavior you're seeing in this case is expected.

Here's the way you can make this work through type annotations:
```
def foo(a: Literal[0, 1], b: Literal[0]):
    my_dict: Dict[Tuple[Literal[0, 1], Literal[0]], str] = {
        (0, 0): "",
        (1, 0): "",
    }
    value = my_dict[(a, b)]
```

**Comment by Congee** [comment]
ummm, is it wrong to annotate the keys of `my_dict` as a union of those literals?
```python
def foo(a: Literal[0, 1], b: Literal[0]):
    my_dict: Dict[
        Union[
            Tuple[Literal[0], Literal[0]],
            Tuple[Literal[1], Literal[0]],
        ],
        str
    ] = {
    # my_dict: Dict[Tuple[Literal[0, 1], Literal[0]], str] = {
        (0, 0): "",
        (1, 0): "",
    }
    _ = my_dict[(a, b)]  # pyright is not happy here
```

**Comment by erictraut** [comment]
The `Dict` type is a mutable container, so it marks its type parameters as invariant. That means the key and value types need to match exactly. In your example above, there is a mismatch.

`a` is of type `Tuple[Literal[0, 1]]`, but the key type for the my_dict must be of type `Union[Tuple[Literal[0], Literal[0]], Tuple[Literal[1], Literal[0]]]`. Because of the invariance, the types must match exactly.

Another issue with your example is that:
`Union[Tuple[Literal[0], Literal[0]], Tuple[Literal[1], Literal[0]]]` is not equivalent to `Tuple[Union[Literal[0], Literal[1]], Literal[0]]`.

I confirmed that mypy also flags this as an error.

**Comment by Congee** [comment]
make sense! thank you for the detailed explanation!

**Comment by erictraut** [comment]
BTW, I made some improvements to the error messages so they are clearer.

**Comment by erictraut** [comment]
I've decided to revisit this issue and change the way that pyright infers tuple expressions. In particular, when a tuple expression is embedded within another container (list, dict, set, or tuple) expression, the type checker will no longer retain literal values for the tuple. This change will be included in the next release of pyright.

**Comment by erictraut** [comment]
This is included in pyright 1.1.365.
